File: ruby/lib/mable_bank/domain/company_account_book.rb
# frozen_string_literal: true

module MableBank
  module Domain
    class CompanyAccountBook
      def initialize(accounts = [])
        @accounts = {}
        @ledger = nil

        accounts.each { |account| add_account(account) }
      end

      def use_ledger(ledger)
        @ledger = ledger
        self
      end

      def add_account(account)
        register(account)
      end

      def balance_for(account_number)
        account = @accounts[account_number]
        return unless account

        snapshot_for(account)
      end

      def transfer(instruction)
        reason_code = evaluate(instruction)
        return failure(instruction, reason_code) if reason_code

        ledger.atomic { apply(instruction) }
        TransferResult.success(instruction)
      end

      def simulate(instruction)
        reason_code = evaluate(instruction)
        return failure(instruction, reason_code) if reason_code

        TransferResult.success(instruction)
      end

      def final_balances
        @accounts.keys.sort_by(&:value).map { |account_number| snapshot_for(@accounts[account_number]) }
      end

      def capture_balance_snapshot
        @accounts.transform_values(&:balance)
      end

      def restore_balance_snapshot(snapshot)
        snapshot.each do |account_number, balance|
          @accounts[account_number].replace_balance!(balance)
        end
      end

      private

      def ledger
        @ledger || NullLedger.new
      end

      def evaluate(instruction)
        source = @accounts[instruction.from_account_number]
        return TransferFailureReason::SOURCE_NOT_FOUND unless source

        destination = @accounts[instruction.to_account_number]
        return TransferFailureReason::DESTINATION_NOT_FOUND unless destination

        return TransferFailureReason::INSUFFICIENT_FUNDS unless source.can_debit?(instruction.amount)

        nil
      end

      def apply(instruction)
        source = @accounts[instruction.from_account_number]
        destination = @accounts[instruction.to_account_number]

        raise InvariantViolationError, 'apply called without successful evaluation' unless source && destination

        source.debit!(instruction.amount)
        destination.credit(instruction.amount)
      end

      def register(account)
        account_number = account.account_number

        raise DuplicateAccountError, "Duplicate account number: #{account_number}" if @accounts.key?(account_number)

        @accounts[account_number] = copy_account(account)
      end

      def copy_account(account)
        account.copy
      end

      def snapshot_for(account)
        AccountBalance.new(account_number: account.account_number, balance: account.balance)
      end

      def failure(instruction, reason_code)
        TransferResult.failure(instruction, reason_code: reason_code)
      end
    end
  end
end

File: nodejs-typescript/src/domain/CompanyAccountBook.ts
import type { Ledger } from './Ledger.js';
import { Account } from './Account.js';
import { AccountBalance } from './AccountBalance.js';
import type { AccountNumber } from './AccountNumber.js';
import { DuplicateAccountError, InvariantViolationError } from './errors.js';
import { NullLedger } from './NullLedger.js';
import type { Money } from './Money.js';
import type { TransferInstruction } from './TransferInstruction.js';
import {
  transferFailure,
  transferSuccess,
  type TransferFailureReason,
  type TransferResult,
} from './TransferResult.js';

export type BalanceSnapshot = ReadonlyMap<AccountNumber, Money>;

export class CompanyAccountBook {
  private readonly accounts = new Map<string, Account>();
  private ledger: Ledger | null = null;

  constructor(accounts: readonly Account[] = []) {
    for (const account of accounts) {
      this.addAccount(account);
    }
  }

  setLedger(ledger: Ledger): void {
    this.ledger = ledger;
  }

  addAccount(account: Account): void {
    const key = account.accountNumber.toString();
    if (this.accounts.has(key)) {
      throw new DuplicateAccountError(`Duplicate account number: ${key}`);
    }
    this.accounts.set(key, this.copyAccount(account));
  }

  private copyAccount(account: Account): Account {
    const AccountClass = account.constructor as typeof Account;
    return new AccountClass(account.accountNumber, account.getBalance());
  }

  transfer(instruction: TransferInstruction): TransferResult {
    const reason = this.evaluate(instruction);
    if (reason !== null) {
      return transferFailure(instruction, reason);
    }

    this.activeLedger.atomic(() => {
      this.apply(instruction);
    });
    return transferSuccess(instruction);
  }

  simulate(instruction: TransferInstruction): TransferResult {
    const reason = this.evaluate(instruction);
    if (reason !== null) {
      return transferFailure(instruction, reason);
    }
    return transferSuccess(instruction);
  }

  balanceFor(accountNumber: AccountNumber): AccountBalance | undefined {
    const account = this.accounts.get(accountNumber.toString());
    if (!account) return undefined;
    return this.snapshotFor(account);
  }

  finalBalances(): AccountBalance[] {
    return [...this.accounts.values()]
      .map((account) => this.snapshotFor(account))
      .sort((left, right) =>
        left.accountNumber.toString().localeCompare(right.accountNumber.toString()),
      );
  }

  captureBalanceSnapshot(): BalanceSnapshot {
    const snapshot = new Map<AccountNumber, Money>();
    for (const account of this.accounts.values()) {
      snapshot.set(account.accountNumber, account.getBalance());
    }
    return snapshot;
  }

  restoreBalanceSnapshot(snapshot: BalanceSnapshot): void {
    for (const [accountNumber, balance] of snapshot) {
      const account = this.accounts.get(accountNumber.toString());
      if (account) {
        account.replaceBalance(balance);
      }
    }
  }

  private get activeLedger(): Ledger {
    return this.ledger ?? new NullLedger();
  }

  private evaluate(instruction: TransferInstruction): TransferFailureReason | null {
    const source = this.accounts.get(instruction.fromAccountNumber.toString());
    if (!source) return 'source_not_found';

    const destination = this.accounts.get(instruction.toAccountNumber.toString());
    if (!destination) return 'destination_not_found';

    if (!source.canDebit(instruction.amount)) return 'insufficient_funds';

    return null;
  }

  private apply(instruction: TransferInstruction): void {
    const source = this.accounts.get(instruction.fromAccountNumber.toString());
    const destination = this.accounts.get(instruction.toAccountNumber.toString());
    if (!source || !destination) {
      throw new InvariantViolationError('apply called without successful evaluation');
    }
    source.debit(instruction.amount);
    destination.credit(instruction.amount);
  }

  private snapshotFor(account: Account): AccountBalance {
    return new AccountBalance(account.accountNumber, account.getBalance());
  }
}
